File: teeverify/extract.go
package teeverify

import (
	"fmt"
	"slices"
	"strings"

	attestpb "github.com/Layr-Labs/go-tpm-tools/proto/attest"
	tpmpb "github.com/Layr-Labs/go-tpm-tools/proto/tpm"
	sabi "github.com/google/go-sev-guest/abi"
)

// ExtractTPMClaims extracts TPM-layer claims from a verified TPM attestation.
// This includes PCRs, GCE metadata, and hardened status.
// For TEE-specific claims, use VerifiedTEEAttestation.ExtractTEEClaims().
// For container claims, use Attestation.ExtractContainerClaims().
func (v *VerifiedTPMAttestation) ExtractTPMClaims(opts ExtractOptions) (*TPMClaims, error) {
	claims := &TPMClaims{
		Platform: v.Platform,
		Hardened: isHardened(v.machineState.GetLinuxKernel().GetCommandLine()),
	}

	pcrs, err := extractPCRs(v.attestation, opts.PCRIndices)
	if err != nil {
		return nil, fmt.Errorf("failed to extract PCRs: %w", err)
	}
	claims.PCRs = pcrs

	if p := v.machineState.GetPlatform(); p != nil {
		if info := p.GetInstanceInfo(); info != nil {
			claims.GCE = &GCEInfo{
				ProjectID:     info.GetProjectId(),
				ProjectNumber: info.GetProjectNumber(),
				Zone:          info.GetZone(),
				InstanceID:    info.GetInstanceId(),
				InstanceName:  info.GetInstanceName(),
			}
		}
	}

	return claims, nil
}

// ExtractTEEClaims extracts TEE-specific claims (TDX or SEV-SNP) from a verified TEE attestation.
func (v *VerifiedTEEAttestation) ExtractTEEClaims() (*TEEClaims, error) {
	claims := &TEEClaims{
		Platform: v.Platform,
	}

	switch v.Platform {
	case PlatformIntelTDX:
		tdxClaims, err := extractTDXClaims(v.attestation)
		if err != nil {
			return nil, fmt.Errorf("failed to extract TDX claims: %w", err)
		}
		claims.TDX = tdxClaims
	case PlatformAMDSevSnp:
		sevClaims, err := extractSevSnpClaims(v.attestation)
		if err != nil {
			return nil, fmt.Errorf("failed to extract SEV-SNP claims: %w", err)
		}
		claims.SevSnp = sevClaims
	default:
		return nil, fmt.Errorf("no TEE claims available for platform %s", v.Platform.PlatformTag())
	}

	return claims, nil
}

func extractPCRs(attestation *attestpb.Attestation, indices []uint32) (map[uint32][32]byte, error) {
	// Validate PCR indices (TPM 2.0 has PCRs 0-23)
	for _, idx := range indices {
		if idx > 23 {
			return nil, fmt.Errorf("invalid PCR index %d (must be 0-23)", idx)
		}
	}

	var sha256PCRs map[uint32][]byte
	for _, quote := range attestation.GetQuotes() {
		if quote.GetPcrs().GetHash() == tpmpb.HashAlgo_SHA256 {
			sha256PCRs = quote.GetPcrs().GetPcrs()
			break
		}
	}
	if sha256PCRs == nil {
		return nil, fmt.Errorf("attestation contains no SHA-256 PCR quotes")
	}

	result := make(map[uint32][32]byte, len(indices))
	for _, idx := range indices {
		val, ok := sha256PCRs[idx]
		if !ok {
			return nil, fmt.Errorf("PCR %d not found in attestation", idx)
		}
		if len(val) != 32 {
			return nil, fmt.Errorf("PCR %d has invalid length: got %d, expected 32", idx, len(val))
		}
		var pcrVal [32]byte
		copy(pcrVal[:], val)
		result[idx] = pcrVal
	}

	return result, nil
}

func extractTDXClaims(attestation *attestpb.Attestation) (*TDXClaims, error) {
	quote := attestation.GetTdxAttestation()
	if quote == nil {
		return nil, fmt.Errorf("TDX attestation is nil")
	}

	quoteBody := quote.GetTdQuoteBody()
	if quoteBody == nil {
		return nil, fmt.Errorf("TDX quote body is nil")
	}

	tdAttrs := quoteBody.GetTdAttributes()
	if len(tdAttrs) < 2 {
		return nil, fmt.Errorf("TD attributes too short: got %d bytes, need at least 2", len(tdAttrs))
	}

	claims := &TDXClaims{
		Attributes: TDAttributes{
			Debug:         tdAttrs[0]&0x01 != 0,
			SeptVEDisable: tdAttrs[0]&0x10 != 0,
			PKS:           tdAttrs[0]&0x40 != 0,
			KL:            tdAttrs[0]&0x80 != 0,
			PerfMon:       tdAttrs[1]&0x01 != 0,
		},
	}

	// Extract TEE TCB SVN (must be exactly 16 bytes if present)
	if tcb := quoteBody.GetTeeTcbSvn(); len(tcb) > 0 {
		if len(tcb) != 16 {
			return nil, fmt.Errorf("invalid TeeTcbSvn length: got %d, expected 16", len(tcb))
		}
		copy(claims.TeeTcbSvn[:], tcb)
	}

	// Extract MRTD (must be exactly 48 bytes if present)
	if mrtd := quoteBody.GetMrTd(); len(mrtd) > 0 {
		if len(mrtd) != 48 {
			return nil, fmt.Errorf("invalid MRTD length: got %d, expected 48", len(mrtd))
		}
		copy(claims.MRTD[:], mrtd)
	}

	// Extract RTMRs (each must be exactly 48 bytes if present)
	if rtmrs := quoteBody.GetRtmrs(); len(rtmrs) > 0 {
		if len(rtmrs) != 4 {
			return nil, fmt.Errorf("invalid RTMR count: got %d, expected 4", len(rtmrs))
		}
		for i, rtmr := range rtmrs {
			if len(rtmr) != 48 {
				return nil, fmt.Errorf("invalid RTMR%d length: got %d, expected 48", i, len(rtmr))
			}
		}
		copy(claims.RTMR0[:], rtmrs[0])
		copy(claims.RTMR1[:], rtmrs[1])
		copy(claims.RTMR2[:], rtmrs[2])
		copy(claims.RTMR3[:], rtmrs[3])
	}

	return claims, nil
}

func extractSevSnpClaims(attestation *attestpb.Attestation) (*SevSnpClaims, error) {
	snpAttestation := attestation.GetSevSnpAttestation()
	if snpAttestation == nil {
		return nil, fmt.Errorf("SEV-SNP attestation is nil")
	}

	report := snpAttestation.GetReport()
	if report == nil {
		return nil, fmt.Errorf("SEV-SNP report is nil")
	}

	guestPolicy, err := sabi.ParseSnpPolicy(report.GetPolicy())
	if err != nil {
		return nil, fmt.Errorf("failed to parse guest policy: %w", err)
	}

	claims := &SevSnpClaims{
		CurrentTcb:   report.GetCurrentTcb(),
		ReportedTcb:  report.GetReportedTcb(),
		CommittedTcb: report.GetCommittedTcb(),
		GuestSvn:     report.GetGuestSvn(),
		Policy: SevSnpPolicy{
			Debug:                guestPolicy.Debug,
			MigrateMA:            guestPolicy.MigrateMA,
			SMT:                  guestPolicy.SMT,
			ABIMinor:             guestPolicy.ABIMinor,
			ABIMajor:             guestPolicy.ABIMajor,
			SingleSocket:         guestPolicy.SingleSocket,
			CipherTextHidingDRAM: guestPolicy.CipherTextHidingDRAM,
		},
	}

	// Extract Measurement (must be exactly 48 bytes if present)
	if m := report.GetMeasurement(); len(m) > 0 {
		if len(m) != 48 {
			return nil, fmt.Errorf("invalid Measurement length: got %d, expected 48", len(m))
		}
		copy(claims.Measurement[:], m)
	}

	// Extract HostData (must be exactly 32 bytes if present)
	if h := report.GetHostData(); len(h) > 0 {
		if len(h) != 32 {
			return nil, fmt.Errorf("invalid HostData length: got %d, expected 32", len(h))
		}
		copy(claims.HostData[:], h)
	}

	return claims, nil
}

func isHardened(cmdline string) bool {
	return slices.Contains(strings.Fields(cmdline), "confidential-space.hardened=true")
}

File: sdk/attest-ts/src/extract.ts
import type { Attestation as AttestationProto } from './proto/layr_attest_pb.js';
import { HashAlgo } from './proto/layr_tpm_pb.js';
import { Platform, parseSevSnpPolicy } from './types.js';
import type {
  VerifiedTPMAttestation,
  VerifiedTEEAttestation,
  ExtractOptions,
  TPMClaims,
  TEEClaims,
  TDXClaims,
  SevSnpClaims,
} from './types.js';

/**
 * Extract TPM-layer claims from a verified TPM attestation.
 * Includes PCRs, GCE metadata, and hardened status.
 *
 * Port of VerifiedTPMAttestation.ExtractTPMClaims from Go.
 */
export function extractTPMClaims(
  verified: VerifiedTPMAttestation,
  opts: ExtractOptions,
): TPMClaims {
  const claims: TPMClaims = {
    platform: verified.platform,
    hardened: isHardened(verified.machineState.linuxKernel?.commandLine ?? ''),
    pcrs: extractPCRs(verified.proto, opts.pcrIndices),
  };

  const platformState = verified.machineState.platform;
  if (platformState?.instanceInfo) {
    const info = platformState.instanceInfo;
    claims.gce = {
      projectId: info.projectId,
      projectNumber: info.projectNumber,
      zone: info.zone,
      instanceId: info.instanceId,
      instanceName: info.instanceName,
    };
  }

  return claims;
}

/**
 * Extract TEE-specific claims (TDX or SEV-SNP) from a verified TEE attestation.
 *
 * Port of VerifiedTEEAttestation.ExtractTEEClaims from Go.
 */
export function extractTEEClaims(
  verified: VerifiedTEEAttestation,
): TEEClaims {
  const claims: TEEClaims = {
    platform: verified.platform,
  };

  switch (verified.platform) {
    case Platform.IntelTDX:
      claims.tdx = extractTDXClaims(verified.proto);
      break;
    case Platform.AMDSevSnp:
      claims.sevSnp = extractSevSnpClaims(verified.proto);
      break;
    default:
      throw new Error(`no TEE claims available for platform ${verified.platform}`);
  }

  return claims;
}

// --- Internal helpers ---

function extractPCRs(attestation: AttestationProto, indices: number[]): Map<number, Uint8Array> {
  for (const idx of indices) {
    if (idx > 23) {
      throw new Error(`invalid PCR index ${idx} (must be 0-23)`);
    }
  }

  let sha256PCRs: { [key: number]: Uint8Array } | undefined;
  for (const quote of attestation.quotes) {
    if (quote.pcrs && quote.pcrs.hash === HashAlgo.SHA256) {
      sha256PCRs = quote.pcrs.pcrs;
      break;
    }
  }
  if (!sha256PCRs) {
    throw new Error('attestation contains no SHA-256 PCR quotes');
  }

  const result = new Map<number, Uint8Array>();
  for (const idx of indices) {
    const val = sha256PCRs[idx];
    if (!val) {
      throw new Error(`PCR ${idx} not found in attestation`);
    }
    if (val.length !== 32) {
      throw new Error(`PCR ${idx} has invalid length: got ${val.length}, expected 32`);
    }
    result.set(idx, val);
  }

  return result;
}

function extractTDXClaims(attestation: AttestationProto): TDXClaims {
  if (attestation.teeAttestation.case !== 'tdxAttestation') {
    throw new Error('TDX attestation is nil');
  }

  const quoteBody = attestation.teeAttestation.value.tdQuoteBody;
  if (!quoteBody) {
    throw new Error('TDX quote body is nil');
  }

  const tdAttrs = quoteBody.tdAttributes;
  if (tdAttrs.length < 2) {
    throw new Error(`TD attributes too short: got ${tdAttrs.length} bytes, need at least 2`);
  }

  const claims: TDXClaims = {
    attributes: {
      debug: (tdAttrs[0] & 0x01) !== 0,
      septVEDisable: (tdAttrs[0] & 0x10) !== 0,
      pks: (tdAttrs[0] & 0x40) !== 0,
      kl: (tdAttrs[0] & 0x80) !== 0,
      perfMon: (tdAttrs[1] & 0x01) !== 0,
    },
    mrtd: new Uint8Array(48),
    rtmr0: new Uint8Array(48),
    rtmr1: new Uint8Array(48),
    rtmr2: new Uint8Array(48),
    rtmr3: new Uint8Array(48),
    teeTcbSvn: new Uint8Array(16),
  };

  // Extract TEE TCB SVN (must be exactly 16 bytes if present)
  if (quoteBody.teeTcbSvn.length > 0) {
    if (quoteBody.teeTcbSvn.length !== 16) {
      throw new Error(`invalid TeeTcbSvn length: got ${quoteBody.teeTcbSvn.length}, expected 16`);
    }
    claims.teeTcbSvn.set(quoteBody.teeTcbSvn);
  }

  // Extract MRTD (must be exactly 48 bytes if present)
  if (quoteBody.mrTd.length > 0) {
    if (quoteBody.mrTd.length !== 48) {
      throw new Error(`invalid MRTD length: got ${quoteBody.mrTd.length}, expected 48`);
    }
    claims.mrtd.set(quoteBody.mrTd);
  }

  // Extract RTMRs (each must be exactly 48 bytes if present)
  if (quoteBody.rtmrs.length > 0) {
    if (quoteBody.rtmrs.length !== 4) {
      throw new Error(`invalid RTMR count: got ${quoteBody.rtmrs.length}, expected 4`);
    }
    for (let i = 0; i < 4; i++) {
      if (quoteBody.rtmrs[i].length !== 48) {
        throw new Error(`invalid RTMR${i} length: got ${quoteBody.rtmrs[i].length}, expected 48`);
      }
    }
    claims.rtmr0.set(quoteBody.rtmrs[0]);
    claims.rtmr1.set(quoteBody.rtmrs[1]);
    claims.rtmr2.set(quoteBody.rtmrs[2]);
    claims.rtmr3.set(quoteBody.rtmrs[3]);
  }

  return claims;
}

function extractSevSnpClaims(attestation: AttestationProto): SevSnpClaims {
  if (attestation.teeAttestation.case !== 'sevSnpAttestation') {
    throw new Error('SEV-SNP attestation is nil');
  }

  const report = attestation.teeAttestation.value.report;
  if (!report) {
    throw new Error('SEV-SNP report is nil');
  }

  const policy = parseSevSnpPolicy(report.policy);

  const claims: SevSnpClaims = {
    currentTcb: report.currentTcb,
    reportedTcb: report.reportedTcb,
    committedTcb: report.committedTcb,
    guestSvn: report.guestSvn,
    policy,
    measurement: new Uint8Array(48),
    hostData: new Uint8Array(32),
  };

  // Extract Measurement (must be exactly 48 bytes if present)
  if (report.measurement.length > 0) {
    if (report.measurement.length !== 48) {
      throw new Error(`invalid Measurement length: got ${report.measurement.length}, expected 48`);
    }
    claims.measurement.set(report.measurement);
  }

  // Extract HostData (must be exactly 32 bytes if present)
  if (report.hostData.length > 0) {
    if (report.hostData.length !== 32) {
      throw new Error(`invalid HostData length: got ${report.hostData.length}, expected 32`);
    }
    claims.hostData.set(report.hostData);
  }

  return claims;
}

function isHardened(cmdline: string): boolean {
  return cmdline.split(/\s+/).includes('confidential-space.hardened=true');
}
